### PR TITLE
Add support for host_rewrite_header

### PR DESCRIPTION
**Pull request by davidjumani**
# Description

Adds support for the `host_rewrite_header` route option.
This configures envoy to rewrite the host header with the value of another header based on the value provided
Ref: https://www.envoyproxy.io/docs/envoy/latest/api-v3/config/route/v3/route_components.proto#envoy-v3-api-field-config-route-v3-routeaction-host-rewrite-header

This also fixes a bug where gloo did not reject invalid envoy header keys
`vs.yaml`
```
        headerManipulation:
          requestHeadersToAdd:
          - header:
              key: "head(er)"
              value: random
```

Previously
```
$ kg apply -f vs.yaml 
virtualservice.gateway.solo.io/default created

$ curl localhost:8080/get
invalid header name: head(er)
```

Now
```
$ kg apply -f vs.yaml 

	* Validating *v1.VirtualService failed: validating *v1.VirtualService name:"default" namespace:"gloo-system": 1 error occurred:
	* failed gloo validation resource reports: 2 errors occurred:
	* invalid resource gloo-system.gateway-proxy
	* Route Error: ProcessingError. Reason: *headers.plugin: 'head(er)' is an invalid HTTP header key. Route Name: gloo-system_default-route-0-matcher-0; Route Error: ProcessingError. Reason: *headers.plugin: 'head(er)' is an invalid HTTP header key. Route Name: gloo-system_default-route-0-matcher-0
```

## API changes
- Adds the `host_rewrite_header` to the route options

## Code changes
- Adds support for the `host_rewrite_header` option in the basic route plugin

## CI changes
- N/A

## Docs changes
- N/A

# Context
https://github.com/solo-io/gloo/issues/9579

## Interesting decisions
 N/A

## Testing steps
Unit tests

# Checklist:

- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works

BOT NOTES: 
resolves https://github.com/solo-io/gloo/issues/9622

### CODE REVIEW

File: projects/gloo/pkg/plugins/basicroute/plugin.go [hunk 1]
```diff
@@ -230,7 +230,10 @@ func applyHostRewrite(ctx context.Context, in *v1.Route, out *envoy_config_route
 		routeAction.Route.HostRewriteSpecifier = &envoy_config_route_v3.RouteAction_AutoHostRewrite{
 			AutoHostRewrite: rewriteType.AutoHostRewrite,
 		}
-
+	case *v1.RouteOptions_HostRewriteHeader:
+		routeAction.Route.HostRewriteSpecifier = &envoy_config_route_v3.RouteAction_HostRewriteHeader{
```
Comment (nfuden) [inline projects/gloo/pkg/plugins/basicroute/plugin.go:234]: We should validate that the header is minimally valid for envoy.

Comment (davidjumani) [reply projects/gloo/pkg/plugins/basicroute/plugin.go:239]: Added validation

File: pkg/utils/api_conversion/type.go [hunk 2]
```diff
@@ -71,13 +72,14 @@ func ToEnvoyHeaderValueOptionList(option []*envoycore_sk.HeaderValueOption, secr
 	return result, nil
 }
 
-// CheckForbiddenCustomHeaders checks whether the custom header is allowed to be modified as per https://www.envoyproxy.io/docs/envoy/latest/configuration/http/http_conn_man/headers#custom-request-response-headers
-func CheckForbiddenCustomHeaders(header envoycore_sk.HeaderValue) error {
+// validateCustomHeaders checks whether the custom header is allowed to be modified as per https://www.envoyproxy.io/docs/envoy/latest/configuration/http/http_conn_man/headers#custom-request-response-headers
+// and validates the whether the header will be accepted by envoy
+func validateCustomHeaders(header envoycore_sk.HeaderValue) error {
```
Comment (davidjumani) [inline pkg/utils/api_conversion/type.go:77]: `vs.yaml`
```
        headerManipulation:
          requestHeadersToAdd:
          - header:
              key: "head(er)"
              value: random
```

Previously
```
kg apply -f vs.yaml 

virtualservice.gateway.solo.io/default created

curl localhost:8080/get
invalid header name: head(er)
```

Now
```
```

File: projects/gloo/pkg/plugins/basicroute/plugin.go [hunk 3]
```diff
@@ -230,7 +231,14 @@ func applyHostRewrite(ctx context.Context, in *v1.Route, out *envoy_config_route
 		routeAction.Route.HostRewriteSpecifier = &envoy_config_route_v3.RouteAction_AutoHostRewrite{
 			AutoHostRewrite: rewriteType.AutoHostRewrite,
 		}
-
+	case *v1.RouteOptions_HostRewriteHeader:
+		err := headers.ValidateHeaderKey(rewriteType.HostRewriteHeader.GetValue())
+		if err != nil {
+			return err
```
Comment (nfuden) [inline projects/gloo/pkg/plugins/basicroute/plugin.go:237]: @npolshakova how would you feel if this was a warning? 
namely a bad header for hosts I think fits into the new warning paradigm right?

File: projects/gloo/pkg/plugins/utils/headers/validate.go [hunk 4]
```diff
@@ -0,0 +1,20 @@
+package headers
+
+import (
+	"fmt"
+	"regexp"
+)
+
+// ValidateHeaderKey checks whether a header is valid based on the RFC and envoy's regex to accept a header key
+func ValidateHeaderKey(key string) error {
```
Comment (nfuden) [inline projects/gloo/pkg/plugins/utils/headers/validate.go:9]: nit: works for now but I would love to have this also deal with extra whitespace minimally. 
I guess we should also link it closer to the c++ code some day but thats less important

**Review by nfuden — APPROVED** [review]
While I would like this to be more robust in the header checking area eventually and would really love a follow up issue but this works for me for now

Edit: i am blind: this does link nicely to the validator in envoy

File: projects/gloo/pkg/plugins/basicroute/plugin.go (hunk 1 above)
Comment (davidjumani) [reply projects/gloo/pkg/plugins/basicroute/plugin.go:239]: So envoy doesn't throw any validation error if the header key is invalid. Maybe this check is unneccessary

File: projects/gloo/pkg/plugins/utils/headers/validate.go (hunk 4 above)
Comment (davidjumani) [reply projects/gloo/pkg/plugins/utils/headers/validate.go:9]: Do you mean trim whitespaces ? This rejects keys with whitespace in them

Comment (davidjumani) [reply projects/gloo/pkg/plugins/utils/headers/validate.go:9]: This has been copied from the [envoy repo](https://github.com/envoyproxy/envoy/blob/b0f4332867267913d9aa80c5c0befda14a00d826/source/common/http/character_set_validation.h#L24-L35) and should validate it the same way envoy does

File: changelog/v1.17.0-rc4/add-host-rewrite-header.yaml [hunk 5]
```diff
@@ -0,0 +1,5 @@
+changelog:
+- type: NEW_FEATURE
+  issueLink: https://github.com/solo-io/gloo/issues/9579
+  resolvesIssue: false
+  description: Adds the `host_rewrite_header` to the route options to allow envoy to swapped the host header with the content of given downstream or custom header
```
Comment (sheidkamp) [inline changelog/v1.17.0-rc4/add-host-rewrite-header.yaml:5]: Can we add this note from the [Envoy docs](https://www.envoyproxy.io/docs/envoy/latest/api-v3/config/route/v3/route_components.proto#envoy-v3-api-field-config-route-v3-routeaction-host-rewrite-header): `Pay attention to the potential security implications of using this option. Provided header must come from trusted source.`

File: pkg/utils/api_conversion/type.go [hunk 6]
```diff
@@ -71,13 +72,14 @@ func ToEnvoyHeaderValueOptionList(option []*envoycore_sk.HeaderValueOption, secr
 	return result, nil
 }
 
-// CheckForbiddenCustomHeaders checks whether the custom header is allowed to be modified as per https://www.envoyproxy.io/docs/envoy/latest/configuration/http/http_conn_man/headers#custom-request-response-headers
-func CheckForbiddenCustomHeaders(header envoycore_sk.HeaderValue) error {
```
Comment (sheidkamp) [inline pkg/utils/api_conversion/type.go:75]: Should we consider this a breaking change due to the removal of an exported function, or continue using the same function name?

File: projects/gloo/pkg/plugins/utils/headers/validate.go (hunk 4 above)
Comment (nfuden) [reply projects/gloo/pkg/plugins/utils/headers/validate.go:9]: oh wow I was blind. i didnt see your " " and I guess i ran it through the old validation in playground. This does seem to work!

**Review by nfuden — APPROVED** [review]

**Review by sheidkamp — COMMENTED** [review]
A couple notes/questions

File: pkg/utils/api_conversion/type.go (hunk 6 above)
Comment (davidjumani) [reply pkg/utils/api_conversion/type.go:75]: I guess I can keep this and wrap this inside the new validation function

File: changelog/v1.17.0-rc4/add-host-rewrite-header.yaml (hunk 5 above)
Comment (davidjumani) [reply changelog/v1.17.0-rc4/add-host-rewrite-header.yaml]: Added

File: projects/gloo/pkg/plugins/basicroute/plugin.go (hunk 1 above)
Comment (nfuden) [reply projects/gloo/pkg/plugins/basicroute/plugin.go:239]: Nah its good practice. Its one of the pieces of value that gloo provides.

File: changelog/v1.18.0-beta1/add-host-rewrite-header.yaml [hunk 7]
```diff
@@ -0,0 +1,5 @@
+changelog:
```
Comment (nfuden) [inline changelog/v1.18.0-beta1/add-host-rewrite-header.yaml:1]: We need to mention the headers validation fix in a user facing way

File: changelog/v1.18.0-beta1/add-host-rewrite-header.yaml [hunk 8]
```diff
@@ -0,0 +1,5 @@
...
+- type: NEW_FEATURE
+  issueLink: https://github.com/solo-io/gloo/issues/9579
+  resolvesIssue: false
+  description: Adds the `host_rewrite_header` to the route options to allow envoy to swapped the host header with the content of given downstream or custom header. Pay attention to the potential security implications of using this option. Provided header must come from trusted source.
```
Comment (nfuden) [inline changelog/v1.18.0-beta1/add-host-rewrite-header.yaml:5]: This does not display well in our changelog. Try out the changelog strategy locally and you will see that this is truncated
```suggestion
  description: >|
    Adds the `host_rewrite_header` to the route options to allow envoy to swapped the host header with the content of given downstream or custom header. Pay attention to the potential security implications of using this option. Provided header must come from trusted source.
```

File: changelog/v1.18.0-beta1/add-host-rewrite-header.yaml (hunk 7 above)
Comment (davidjumani) [reply changelog/v1.18.0-beta1/add-host-rewrite-header.yaml:1]: Added

File: changelog/v1.18.0-beta1/add-host-rewrite-header.yaml [hunk 9]
```diff
@@ -0,0 +1,11 @@
...
+- type: NEW_FEATURE
+  issueLink: https://github.com/solo-io/gloo/issues/9579
+  resolvesIssue: false
+  description: >-
+    Adds the `host_rewrite_header` to the route options to allow envoy to swapped the host header with the content of given downstream or custom header. Pay attention to the potential security implications of using this option. Provided header must come from trusted source.
+- type: FIX
+  issueLink: https://github.com/solo-io/gloo/issues/9622
+  resolvesIssue: true
+  description: >-
+    Header names are now validated according to RFC 9110, https://www.rfc-editor.org/rfc/rfc9110.html#section-5.1. If an invalid header name is passed, it is rejected
```
Comment (nfuden) [inline changelog/v1.18.0-beta1/add-host-rewrite-header.yaml:11]: Can we make this a bit more user facing?
What headers were we validating. 
What was the gap that we are now catching?

File: projects/gloo/pkg/plugins/utils/headers/validate.go [hunk 10]
```diff
@@ -0,0 +1,20 @@
...
+	if len(key) == 0 {
+		return fmt.Errorf("empty HTTP header names are not allowed")
+	}
+	// check that header name is only valid ASCII characters
+	// https://github.com/envoyproxy/envoy/blob/b0f4332867267913d9aa80c5c0befda14a00d826/source/common/http/character_set_validation.h#L24-L35
+	validHeaderRegex := regexp.MustCompile("^([a-zA-Z0-9!#$%&'*+.^_`|~-])+$")
```
Comment (nfuden) [inline projects/gloo/pkg/plugins/utils/headers/validate.go:15]: Should this be in the path of translation? or should it be a once or something?

Comment (davidjumani) [reply projects/gloo/pkg/plugins/utils/headers/validate.go]: Good idea. Moved it outside

File: changelog/v1.18.0-beta1/add-host-rewrite-header.yaml (hunk 9 above)
Comment (davidjumani) [reply changelog/v1.18.0-beta1/add-host-rewrite-header.yaml]: Updated with a more descriptive changelog

**Review by sheidkamp — APPROVED** [review]

**Review by nfuden — APPROVED** [review]

**Comment by davidjumani** [comment]
kick bulldozer

**Pull request merged**